src/librustdoc/html/highlight.rs

```diff
@@ -151,151 +150,386 @@ fn can_merge(class1: Option<Class>, class2: Option<Class>, text: &str) -> bool {
     }
 }
 
+#[derive(Debug)]
+struct Content {
+    text: String,
+    /// If `Some` and the `span` is different from the parent, then it might generate a link so we
+    /// need to keep this information.
+    class: Option<Class>,
+    needs_escape: bool,
+}
+
+#[derive(Debug)]
+struct Element {
+    /// If `class` is `None`, then it's just plain text with no HTML tag.
+    class: Option<Class>,
+    /// Content for the current element.
+    content: Vec<Content>,
+}
+
+impl Element {
+    fn new(class: Option<Class>, text: String, needs_escape: bool) -> Self {
+        Self { class, content: vec![Content { text, class, needs_escape }] }
+    }
+
+    fn can_merge(&self, other: &Self) -> bool {
+        other.content.iter().all(|c| can_merge(self.class, other.class, &c.text))
+    }
+
+    fn write_elem_to<W: Write>(
+        &self,
+        out: &mut W,
+        href_context: &Option<HrefContext<'_, '_>>,
+        parent_class: Option<Class>,
+    ) {
+        let mut prev = parent_class;
+        let mut closing_tag = None;
+        for part in &self.content {
+            let text: &dyn Display =
+                if part.needs_escape { &EscapeBodyText(&part.text) } else { &part.text };
```

Comment:
FYI, [`Either` impls `Display`](https://docs.rs/either/latest/either/enum.Either.html#impl-Display-for-Either%3CL,+R%3E), which can be nicer than a `dyn` ref (and maybe slightly more performant)